text/go/0001-go-bom.md

```diff
@@ -0,0 +1,163 @@
+CycloneDX supported-tool that creates valid CycloneDX BOMs for Go applications. CycloneDX is a BOM format supported by
+the OWASP Foundation, and is a serious contender (at the time of this RFC creation) as a format we will eventually
+support.
+
+The CycloneDX Go Module tool is available as a pre-built binary for multiple platforms, and also it can easily be
+installed via `go install github.com/CycloneDX/cyclonedx-gomod`.
+
+This tool supports three different cases of BOM generation:
+
+1. `app`: Include only those modules that the target application actually depends on. Modules required by tests or
+   packages that are not imported by the application are not included. Build constraints are evaluated, which enables a
+   very detailed view of what's really compiled into an application's binary.
+2. `mod`: Include the aggregate of modules required by all packages in the target module. This optionally includes
+   modules required by tests and test packages. Build constraints are NOT evaluated, allowing for a "whole picture" view
+   on the target module's dependencies.
+3. `bin`: Offers support of generating rudimentary SBOMs from binaries built with Go modules.
+
+Of these, we are only interested in `app`. The buildpacks do not currently support building modules, or libraries, which
+is the primary use-case for `mod`. The `bin` command is for pre-built executables, which is unnecessary for the
+buildpack as it has access to the source code. The BOM created by the `bin` command also contains less information than
+the one created via the `app` command.
+
+The output BOM contains the following fields for each module:
+
+* name,
+* version,
+* hash,
+* source URI,
+* license ID,
+* and package URL
+
+Note: The `license` field can be retrieved when the flag `-licenses` is passed to the command.
+
+Almost all of fields that we aim to support
+per [the Paketo BOM RFC](https://github.com/paketo-buildpacks/rfcs/blob/main/text/0033-bill-of-materials.md#overall-schema)
+are available. The only one that is not available is `description`, and this is because there is no standard way to
+capture this information in Go projects (as compared with other languages like nodejs and rust, where the package
+description files contain a field for this metadata). See unresolved questions below.
+
+In order to support this tool in both the online and offline cases, we will host the binary tool on
+the [dep-server](https://github.com/paketo-buildpacks/dep-server).
+
+### Performance
+
+The addition of this tool to the language family order groupings might have an impactful change to the overall
+performance of the buildpacks.
+
+Performance metrics are still TBD, but for a typical project (e.g. `pack`) we observe that it takes about 10-30 seconds
+to construct a BOM.
+
+We can investigate ways to improve this performance.
+
+### Language Family Additions
```

Comment:
@sophiewigmore was the analogous pre-populated `node_modules` case handled in the Nodejs BOM buildpack? What did y'all decide to do?

Reply:
@fg-j Sorry, I thought your previous comment was asking what the desired behavior is in the case where there is a populated `vendor` directory but no `go.mod` file. In that case `cyclonedx-gomod` fails to execute, so we propose that we do not support that case.

In the case where there is a valid `go.mod` file _and_ a pre-populated `vendor` directory (and vendoring is enabled), then if the `go.mod` file and the `vendor` directory are consistent, this is supported with some limitations. According to the `cyclonedx-gomod` README:

>Modules that use vendoring are, although in a limited manner, supported.
Limitations are as follows:
>
>* **No hashes.** Go doesn't copy all module files to vendor, only those that are required to build and test the main module. Because module checksums consider almost all files in a module's directory though, calculating accurate hashes from the vendor directory is not possible. As a consequence, SBOMs for modules that use vendoring do not include component hashes.
>* **License detection may fail.** Go doesn't always copy license files when vendoring modules, which may cause license detection to fail.

If the `go.mod` file is inconsistent with the `vendor` directory, is this even a valid situation for a go project?

Am I addressing your question, or am I still missing something?

Reply:
and yes, in the `node` case, the `cyclonedx-node-module` supports the pre-populated vendor directory without a `package.json`. That is essentially a difference between the `cyclonedx-node-module` and `cyclonedx-gomod` tools.

Reply:
> If the go.mod file is inconsistent with the vendor directory, is this even a valid situation for a go project?

If you're asking "Will the Go buildpack build it, as it's currently implemented?", I'm pretty sure the answer is yes. If you're asking "Is this a well-formed app?", IMO the answer is no.